File: java/fory-core/src/test/java/org/apache/fory/type/Float16Test.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.fory.type;

import static org.testng.Assert.*;

import org.testng.annotations.Test;

/** 测试 Float16 类的 IEEE 754 binary16 实现。 */
public class Float16Test {

  @Test
  public void testSpecialValues() {
    // NaN
    assertTrue(Float16.NaN.isNaN());
    assertFalse(Float16.NaN.equalsValue(Float16.NaN)); // NaN != NaN
    assertFalse(Float16.NaN.isFinite());
    assertFalse(Float16.NaN.isInfinite());

    // Infinity
    assertTrue(Float16.POSITIVE_INFINITY.isInfinite());
    assertTrue(Float16.NEGATIVE_INFINITY.isInfinite());
    assertFalse(Float16.POSITIVE_INFINITY.signbit());
    assertTrue(Float16.NEGATIVE_INFINITY.signbit());
    assertFalse(Float16.POSITIVE_INFINITY.isFinite());

    // Zero
    assertTrue(Float16.ZERO.isZero());
    assertTrue(Float16.NEGATIVE_ZERO.isZero());
    assertTrue(Float16.ZERO.equalsValue(Float16.NEGATIVE_ZERO)); // +0 == -0
    assertFalse(Float16.ZERO.signbit());
    assertTrue(Float16.NEGATIVE_ZERO.signbit());

    // One
    assertFalse(Float16.ONE.isZero());
    assertTrue(Float16.ONE.isNormal());
    assertEquals(1.0f, Float16.ONE.floatValue(), 0.0f);
  }

  @Test
  public void testConversionBasic() {
    // 零值
    assertEquals((short) 0x0000, Float16.valueOf(0.0f).toBits());
    assertEquals((short) 0x8000, Float16.valueOf(-0.0f).toBits());

    // 正常值
    assertEquals((short) 0x3c00, Float16.valueOf(1.0f).toBits()); // 1.0
    assertEquals((short) 0xbc00, Float16.valueOf(-1.0f).toBits()); // -1.0
    assertEquals((short) 0x4000, Float16.valueOf(2.0f).toBits()); // 2.0

    // 往返转换
    assertEquals(1.0f, Float16.valueOf(1.0f).floatValue(), 0.0f);
    assertEquals(-1.0f, Float16.valueOf(-1.0f).floatValue(), 0.0f);
    assertEquals(2.0f, Float16.valueOf(2.0f).floatValue(), 0.0f);
  }

  @Test
  public void testBoundaryValues() {
    // 最大值 65504
    Float16 max = Float16.valueOf(65504.0f);
    assertEquals(0x7bff, max.toBits());
    assertEquals(65504.0f, max.floatValue(), 0.0f);
    assertTrue(max.isNormal());
    assertTrue(max.isFinite());

    // 最小正规数 2^-14
    Float16 minNormal = Float16.valueOf((float) Math.pow(2, -14));
    assertEquals(0x0400, minNormal.toBits());
    assertTrue(minNormal.isNormal());
    assertFalse(minNormal.isSubnormal());

    // 最小次正规数 2^-24
    Float16 minSubnormal = Float16.valueOf((float) Math.pow(2, -24));
    assertEquals(0x0001, minSubnormal.toBits());
    assertTrue(minSubnormal.isSubnormal());
    assertFalse(minSubnormal.isNormal());
  }

  @Test
  public void testOverflowUnderflow() {
    // 上溢到 Inf
    Float16 overflow = Float16.valueOf(70000.0f);
    assertTrue(overflow.isInfinite());
    assertFalse(overflow.signbit());

    Float16 negOverflow = Float16.valueOf(-70000.0f);
    assertTrue(negOverflow.isInfinite());
    assertTrue(negOverflow.signbit());

    // 下溢到零
    Float16 underflow = Float16.valueOf((float) Math.pow(2, -30));
    assertTrue(underflow.isZero());
  }

  @Test
  public void testRoundingToNearestEven() {
    // 测试 round-to-nearest-even 舍入模式
    // 1.0009765625 在 float16 中应该舍入到 1.0
    Float16 rounded = Float16.valueOf(1.0009765625f);
    // Float16 的精度有限，1.0009765625 会舍入到最接近的可表示值
    // 实际上 float16 的精度约为 0.001，所以这个值会保留一些精度
    assertTrue(Math.abs(rounded.floatValue() - 1.0009765625f) < 0.01f);

    // 测试 ties-to-even：恰好在两个 float16 值中间时，舍入到偶数
    // 这需要精确的测试用例，暂时跳过详细测试
  }

  @Test
  public void testArithmetic() {
    Float16 one = Float16.ONE;
    Float16 two = Float16.valueOf(2.0f);
    Float16 three = Float16.valueOf(3.0f);

    // 加法
    assertEquals(3.0f, one.add(two).floatValue(), 0.01f);
    assertEquals(5.0f, two.add(three).floatValue(), 0.01f);

    // 减法
    assertEquals(-1.0f, one.subtract(two).floatValue(), 0.01f);
    assertEquals(1.0f, three.subtract(two).floatValue(), 0.01f);

    // 乘法
    assertEquals(2.0f, one.multiply(two).floatValue(), 0.01f);
    assertEquals(6.0f, two.multiply(three).floatValue(), 0.01f);

    // 除法
    assertEquals(0.5f, one.divide(two).floatValue(), 0.01f);
    assertEquals(1.5f, three.divide(two).floatValue(), 0.01f);

    // 取反
    assertEquals(-1.0f, one.negate().floatValue(), 0.0f);
    assertEquals(1.0f, one.negate().negate().floatValue(), 0.0f);

    // 绝对值
    assertEquals(1.0f, one.abs().floatValue(), 0.0f);
    assertEquals(1.0f, one.negate().abs().floatValue(), 0.0f);
  }

  @Test
  public void testComparison() {
    Float16 one = Float16.ONE;
    Float16 two = Float16.valueOf(2.0f);
    Float16 nan = Float16.NaN;

    // compareTo
    assertTrue(one.compareTo(two) < 0);
    assertTrue(two.compareTo(one) > 0);
    assertEquals(0, one.compareTo(Float16.ONE));

    // equalsValue (IEEE 754 语义)
    assertTrue(one.equalsValue(Float16.ONE));
    assertFalse(nan.equalsValue(nan)); // NaN != NaN
    assertTrue(Float16.ZERO.equalsValue(Float16.NEGATIVE_ZERO)); // +0 == -0

    // equals (位模式相等)
    assertTrue(one.equals(Float16.ONE));
    assertFalse(Float16.ZERO.equals(Float16.NEGATIVE_ZERO)); // 位模式不同
  }

  @Test
  public void testClassification() {
    // isNormal
    assertTrue(Float16.ONE.isNormal());
    assertFalse(Float16.ZERO.isNormal());
    assertFalse(Float16.NaN.isNormal());
    assertFalse(Float16.POSITIVE_INFINITY.isNormal());

    // isSubnormal
    Float16 subnormal = Float16.valueOf((float) Math.pow(2, -24));
    assertTrue(subnormal.isSubnormal());
    assertFalse(Float16.ONE.isSubnormal());

    // isFinite
    assertTrue(Float16.ONE.isFinite());
    assertTrue(Float16.ZERO.isFinite());
    assertFalse(Float16.NaN.isFinite());
    assertFalse(Float16.POSITIVE_INFINITY.isFinite());

    // signbit
    assertFalse(Float16.ONE.signbit());
    assertTrue(Float16.valueOf(-1.0f).signbit());
    assertFalse(Float16.ZERO.signbit());
    assertTrue(Float16.NEGATIVE_ZERO.signbit());
  }

  @Test
  public void testToString() {
    assertEquals("1.0", Float16.ONE.toString());
    assertEquals("2.0", Float16.valueOf(2.0f).toString());
    assertTrue(Float16.NaN.toString().contains("NaN"));
    assertTrue(Float16.POSITIVE_INFINITY.toString().contains("Infinity"));
  }

  @Test
  public void testHashCode() {
    // 相同值应该有相同的 hashCode
    assertEquals(Float16.ONE.hashCode(), Float16.valueOf(1.0f).hashCode());

    // 不同值应该有不同的 hashCode（大概率）
    assertNotEquals(Float16.ONE.hashCode(), Float16.valueOf(2.0f).hashCode());
  }

  @Test
  public void testNumberConversions() {
    Float16 f16 = Float16.valueOf(3.14f);

    // floatValue
    assertEquals(3.14f, f16.floatValue(), 0.01f);

    // doubleValue
    assertEquals(3.14, f16.doubleValue(), 0.01);

    // intValue
    assertEquals(3, f16.intValue());

    // longValue
    assertEquals(3L, f16.longValue());

    // byteValue
    assertEquals((byte) 3, f16.byteValue());

    // shortValue
    assertEquals((short) 3, f16.shortValue());
  }

  @Test
  public void testAllBitPatterns() {
    // 测试所有 65536 个可能的位模式
    // 验证非 NaN 值可以精确往返转换
    int nanCount = 0;
    int normalCount = 0;
    int subnormalCount = 0;
    int zeroCount = 0;
    int infCount = 0;

    for (int bits = 0; bits <= 0xFFFF; bits++) {
      Float16 h = Float16.fromBits((short) bits);
      float f = h.floatValue();
      Float16 h2 = Float16.valueOf(f);

      if (h.isNaN()) {
        nanCount++;
        // NaN 应该保持为 NaN
        assertTrue(h2.isNaN(), "NaN should remain NaN after round-trip");
      } else {
        // 非 NaN 值应该精确往返（或者至少保持相同的值）
        if (!h.equals(h2)) {
          // 允许一些舍入误差，但值应该接近
          assertEquals(
              h.floatValue(),
              h2.floatValue(),
              0.0001f,
              String.format("Round-trip failed for bits 0x%04x", bits));
        }

        if (h.isNormal()) {
          normalCount++;
        } else if (h.isSubnormal()) {
          subnormalCount++;
        } else if (h.isZero()) {
          zeroCount++;
        } else if (h.isInfinite()) {
          infCount++;
        }
      }
    }

    // 验证统计信息合理
    assertTrue(nanCount > 0, "Should have NaN values");
    assertTrue(normalCount > 0, "Should have normal values");
    assertTrue(subnormalCount > 0, "Should have subnormal values");
    assertEquals(2, zeroCount, "Should have exactly 2 zeros (+0 and -0)");
    assertEquals(2, infCount, "Should have exactly 2 infinities (+Inf and -Inf)");

    // 总数应该是 65536
    int total = nanCount + normalCount + subnormalCount + zeroCount + infCount;
    assertEquals(65536, total, "Total should be 65536");
  }

  @Test
  public void testConstants() {
    // 验证常量的位模式
    assertEquals((short) 0x7e00, Float16.NaN.toBits());
    assertEquals((short) 0x7c00, Float16.POSITIVE_INFINITY.toBits());
    assertEquals((short) 0xfc00, Float16.NEGATIVE_INFINITY.toBits());
    assertEquals((short) 0x0000, Float16.ZERO.toBits());
    assertEquals((short) 0x8000, Float16.NEGATIVE_ZERO.toBits());
    assertEquals((short) 0x3c00, Float16.ONE.toBits());
    assertEquals((short) 0x7bff, Float16.MAX_VALUE.toBits());
    assertEquals((short) 0x0400, Float16.MIN_NORMAL.toBits());
    assertEquals((short) 0x0001, Float16.MIN_VALUE.toBits());
  }

  @Test
  public void testSizeConstants() {
    assertEquals(16, Float16.SIZE_BITS);
    assertEquals(2, Float16.SIZE_BYTES);
  }
}

File: dart/packages/fory-test/test/datatype_test/float16_test.dart
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

// @Skip()
import 'dart:math' as math;
import 'package:test/test.dart';
import 'package:fory/src/datatype/float16.dart';
import 'package:fory/src/memory/byte_writer.dart';
import 'package:fory/src/memory/byte_reader.dart';

void main() {
  group('Float16 Conversion', () {
    test('Zero', () {
      expect(Float16.positiveZero.toDouble(), 0.0);
      expect(Float16.negativeZero.toDouble(), -0.0);
      expect(Float16.fromDouble(0.0).toBits(), 0x0000);
      expect(Float16.fromDouble(-0.0).toBits(), 0x8000);
      
      // Verify -0.0 vs 0.0 distinction
      expect(Float16.fromDouble(-0.0).isNegative, true);
      expect(Float16.fromDouble(0.0).isNegative, false);
    });

    test('Infinity', () {
      expect(Float16.positiveInfinity.toDouble(), double.infinity);
      expect(Float16.negativeInfinity.toDouble(), double.negativeInfinity);
      expect(Float16.fromDouble(double.infinity).toBits(), 0x7C00);
      expect(Float16.fromDouble(double.negativeInfinity).toBits(), 0xFC00);
    });

    test('NaN', () {
      expect(Float16.nan.toDouble().isNaN, true);
      expect(Float16.fromDouble(double.nan).isNaN, true);
      // Canonical NaN
      expect(Float16.fromDouble(double.nan).toBits(), 0x7E00);
    });

    test('Exact values', () {
      expect(Float16(1.0).toDouble(), 1.0);
      expect(Float16(-1.0).toDouble(), -1.0);
      expect(Float16(1.5).toDouble(), 1.5);
      expect(Float16(0.5).toDouble(), 0.5);
      expect(Float16(65504.0).toDouble(), 65504.0); // Max Value
    });

    test('Rounding', () {
      // 1.0 is 0x3C00. Next is 1 + 2^-10 = 1.0009765625 (0x3C01)
      // Halfway is 1.00048828125
      
      // Round down
      expect(Float16.fromDouble(1.0004).toDouble(), 1.0);
      
      // Round up
      expect(Float16.fromDouble(1.0006).toDouble(), 1.0009765625);
      
      // Tie to even (1.0 is even (last bit 0))
      // 1.0 + half_epsilon = 1.00048828125
      // 3C00 is even. 3C01 is odd. 
      // If result is exactly halfway, pick even.
      // 1.0 corresponds to stored bits ...00
      
      double one = 1.0;
      double next = 1.0009765625;
      double mid = (one + next) / 2;
      
      // 1.0 (0x3C00) is even. 
      expect(Float16.fromDouble(mid).toBits(), 0x3C00); // 1.0
      
      // 1.0009765625 (0x3C01) is odd.
      // Next is 1.001953125 (0x3C02) - even.
      // Mid between 3C01 and 3C02
      double val1 = 1.0009765625;
      double val2 = 1.001953125;
      double mid2 = (val1 + val2) / 2;
      expect(Float16.fromDouble(mid2).toBits(), 0x3C02); // Round to even (up)
    });

    test('Subnormal', () {
      // Min subnormal: 2^-24 approx 5.96e-8
      double minSub = math.pow(2, -24).toDouble();
      expect(Float16.fromDouble(minSub).toDouble(), minSub);
      expect(Float16.fromDouble(minSub).toBits(), 0x0001);
      
      // Below min subnormal -> 0
      expect(Float16.fromDouble(minSub / 2.1).toBits(), 0x0000);
      // Round up to min subnormal
      expect(Float16.fromDouble(minSub * 0.6).toBits(), 0x0001);
      
      // Max subnormal: 0x03FF = (1 - 2^-10) * 2^-14
      int maxSubBits = 0x03FF;
      Float16 maxSub = Float16.fromBits(maxSubBits);
      expect(maxSub.isSubnormal, true);
      expect(Float16.fromDouble(maxSub.toDouble()).toBits(), maxSubBits);
    });
  });

  group('Float16 Arithmetic', () {
    test('Add', () {
      expect((Float16(1.0) + Float16(2.0)).toDouble(), 3.0);
      expect((Float16(1.0) + 2.0).toDouble(), 3.0);
    });
    test('Sub', () {
      expect((Float16(3.0) - Float16(1.0)).toDouble(), 2.0);
    });
    test('Mul', () {
      expect((Float16(2.0) * Float16(3.0)).toDouble(), 6.0);
    });
    test('Div', () {
      expect((Float16(1.0) / Float16(2.0)), 0.5);
    });
    test('Neg', () {
      expect((-Float16(1.0)).toDouble(), -1.0);
    });
    test('Abs', () {
      expect(Float16(-1.0).abs().toDouble(), 1.0);
    });
  });
  
  group('Float16 Comparision', () {
    test('Equality', () {
       expect(Float16(1.0) == Float16(1.0), true);
       expect(Float16(1.0) == Float16(2.0), false);
       expect(Float16.nan == Float16.nan, true); // Bitwise equality
    });
    
    test('Less/Greater', () {
       expect(Float16(1.0) < Float16(2.0), true);
       expect(Float16(2.0) > Float16(1.0), true);
    });
  });

  group('Serialization', () {
    test('RoundTrip', () {
      var bw = ByteWriter();
      var f1 = Float16(1.234375); // 0x3D3C approx
      var f2 = Float16(-100.0);
      var f3 = Float16.nan;
      var f4 = Float16.positiveInfinity;
      
      bw.writeFloat16(f1);
      bw.writeFloat16(f2);
      bw.writeFloat16(f3);
      bw.writeFloat16(f4);
      
      var bytes = bw.toBytes();
      var br = ByteReader.forBytes(bytes);
      
      var r1 = br.readFloat16();
      var r2 = br.readFloat16();
      var r3 = br.readFloat16();
      var r4 = br.readFloat16();
      
      expect(r1.toDouble(), f1.toDouble());
      expect(r2.toDouble(), f2.toDouble());
      expect(r3.isNaN, true);
      expect(r4.isInfinite, true);
      expect(r4.sign, 1);
    });
  });
}
